src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java

```diff
@@ -377,12 +386,41 @@ private static <T> T checkNull(T obj) {
                 System.out.println(p11Info);
             }
 
-            if ((slotID < 0) || showInfo) {
+            if ((slotID < 0) || showInfo|| (tokenLabel != null)) {
                 long[] slots = p11.C_GetSlotList(false);
                 if (showInfo) {
-                    System.out.println("All slots: " + toString(slots));
-                    slots = p11.C_GetSlotList(true);
-                    System.out.println("Slots with tokens: " + toString(slots));
+                    if (isZ) {
+                        System.out.println("Slots[slotID, tokenName]:");
+                        for (int i = 0; i < slots.length; i++) {
+                            System.out.println("[" + i + ", " + new String(p11.C_GetTokenInfo(slots[i]).label).trim() + "]");
+                        }
+                    } else {
+                        System.out.println("All slots: " + toString(slots));
+                        long[] slotsWithTokens = p11.C_GetSlotList(true);
+                        System.out.println("Slots with tokens: " + toString(slotsWithTokens));
```

Comment:
Please minimize the change relative to upstream, i.e. don't introduce a new local variable:
```java
                    slots = p11.C_GetSlotList(true);
                    System.out.println("Slots with tokens: " + toString(slots));
```
The use of `slots` below for Z will be unaffected.